UseCases/Sources/InstrumentsAppTracing/OSSignpostTracing.swift

```diff
@@ -0,0 +1,267 @@
+        // // TODO: is this right or we should rely on explicit link calls?
+        // if context.signpostID != nil {
+        //     self.addLink(SpanLink(context: context))
+        // }
+
+        // replace signpostID with "us" i.e. this span
+        let signpostID = OSSignpostID(
+            log: log,
+            object: self
+        )
+        self.context.signpostID = signpostID
+
+        if self.isRecording {
+            os_signpost(
+                .begin,
+                log: self.log,
+                name: self.signpostName,
+                signpostID: self.signpostID,
+                Self.beginFormat,
+                self.signpostID.rawValue,
+                "\(context.signpostTraceParentIDs.map({ "\($0.rawValue)" }).joined(separator: ","))",
+                operationName
+            )
+        }
+    }
+
+    #if DEBUG
+    deinit {
+        // sanity checking if we don't accidentally drop spans on the floor without ending them
+        self.lock.lock() // TODO: somewhat bad idea, we should rather implement endTimestamp as an atomic that's lockless to read (!)
+        defer { self.lock.lock() }
+        if self.endTimestamp == nil {
+            print("""
+                  warning: 
+                  Span \(self.signpostID) (\(self.operationName)) \
+                  [todo:source location] \
+                  was dropped without end() being called!
+                  """)
+        }
+    }
+    #endif
+
+
+    public func addLink(_ link: SpanLink) {
+        guard self.isRecording else { return }
+        self.lock.lock()
+        defer { self.lock.unlock() }
+
+        guard let id = link.context.signpostID else {
+            print(
+                """
+                Attempted to addLink(\(link)) to \(self.signpostID) (\(self.operationName))\
+                but no `signpostID` present in passed in baggage context!
+                """
+            )
+            return
+        }
+
+        self.context.signpostTraceParentIDs += [id]
+    }
+
+    public func addEvent(_ event: SpanEvent) {
+        guard self.isRecording else { return }
+
+        // perhaps emit it as os_signpost(.event, ...)
+    }
+
+    public var attributes: SpanAttributes {
+        get {
+            [:] // ignore
+        }
+        set {
+            // ignore
+        }
+    }
+
+
+    public func end(at timestamp: Timestamp) {
+        guard self.isRecording else { return }
+        self.lock.lock()
+        defer { self.lock.unlock() }
+
+        guard self.endTimestamp == nil else {
+            print("warning: attempted to end() more-than-once the span: \(self.signpostID) (\(self.operationName))!")
+            return
+        }
+        self.endTimestamp = timestamp
+
+        os_signpost(
+            .end,
+            log: self.log,
+            name: self.signpostName,
+            signpostID: self.signpostID,
+            Self.endFormat
+        )
```

Comment:
Instruments correlates start/end by the signpost ID, which is why we created it with `self` so we use the objects identity as the correlation ID.